projects/core/src/cms/services/cms-structure-config.service.ts

```diff
@@ -104,32 +119,65 @@ export abstract class CmsStructureConfigService {
     pageStructure: CmsStructureModel,
     slots?: CmsPageSlotsConfig
   ): Observable<CmsStructureModel> {
+    // if no slots have been given, we use the global configured slots
     if (
       !slots &&
       this.cmsDataConfig.cmsStructure &&
       this.cmsDataConfig.cmsStructure.slots
     ) {
       slots = this.cmsDataConfig.cmsStructure.slots;
     }
-    if (slots) {
-      for (const position of Object.keys(slots)) {
-        if (Object.keys(pageStructure.page.slots).indexOf(position) === -1) {
-          // add slot
-          pageStructure.page.slots[position] = slots[position];
-          pageStructure.page.slots[position].uid = position;
 
-          // add slot components
+    if (!slots) {
+      return of(pageStructure);
+    }
+
+    for (const position of Object.keys(slots)) {
+      if (Object.keys(pageStructure.page.slots).indexOf(position) === -1) {
+        // the global slot isn't yet part of the page structure
+        pageStructure.page.slots[position] = {
+          uid: position,
+        };
+
+        for (const c of this.getComponentsByPosition(slots, position)) {
+          if (!pageStructure.page.slots[position].components) {
+            pageStructure.page.slots[position].components = [];
+          }
+          pageStructure.page.slots[position].components.push({
+            uid: c.uid,
+            flexType: c.flexType,
+            typeCode: c.typeCode,
+          });
           if (!pageStructure.components) {
             pageStructure.components = [];
           }
-          if (slots[position].components) {
-            slots[position].components.forEach(c => {
-              pageStructure.components.push(c);
-            });
-          }
+
+          pageStructure.components.push(c);
         }
       }
     }
+
     return of(pageStructure);
   }
+
+  private getComponentsByPosition(
+    slots: CmsPageSlotsConfig,
+    position: string
+  ): ContentSlotComponentData[] {
+    const components = [];
+    if (slots[position] && slots[position].componentIds) {
+      for (const componentId of slots[position].componentIds) {
+        if (
+          this.cmsDataConfig.cmsStructure &&
+          this.cmsDataConfig.cmsStructure.components
+        ) {
+          const c = this.cmsDataConfig.cmsStructure.components[componentId];
```

Comment:
As above, usually using more that one letter to name variables makes them more readable